### PR TITLE
ENH: `_lib`: JAX support (non-jitted)

### PR DESCRIPTION
**Pull request by crusaderky**
- Bump array-api-compat to 1.10 (needed for https://github.com/data-apis/array-api-compat/pull/205)
- Bump array-api-extra to 0.5.0 (needed for https://github.com/data-apis/array-api-extra/pull/53)
- Support JAX in `_lib` and all its unit tests. This doesn't include `@jax.jit` yet (read comments below).

### CODE REVIEW

File: scipy/_lib/_array_api_compat_vendor.py [hunk 1]
```diff
@@ -3,7 +3,9 @@
 # to override functions of array_api_compat.
 
 from .array_api_compat import *  # noqa: F403
-from ._array_api import array_namespace as scipy_array_namespace
 
 # overrides array_api_compat.array_namespace inside array-api-extra
-array_namespace = scipy_array_namespace  # type: ignore[assignment]
+def array_namespace(*args, **kwargs):
+    from ._array_api import array_namespace as scipy_array_namespace
```
Comment (crusaderky) [inline scipy/_lib/_array_api_compat_vendor.py:9]: Prevent circular import introduced by change in _array_api.py above

File: scipy/_lib/_util.py [hunk 2]
```diff
@@ -153,9 +153,9 @@ def _lazywhere(cond, arrays, f, fillvalue=None, f2=None):
         temp2 = xp.asarray(f2(*(arr[ncond] for arr in arrays)))
         dtype = xp.result_type(temp1, temp2)
         out = xp.empty(cond.shape, dtype=dtype)
-        out[ncond] = temp2
+        out = xp_at(out, ncond).set(temp2, copy=None)
 
-    out[cond] = temp1
+    out = xp_at(out, cond).set(temp1, copy=None)
```
Comment (crusaderky) [inline scipy/_lib/_util.py:158]: This won't work under `@jax.jit` because jax doesn't support unknown shapes.
There are two ways forward:
1. (easy-ish?) Reimplement this function in JAX or in array-api-extra and use two different code paths if it is being jitted or not. (_does such a flag exist today?_)
If jitted, use `out = jnp.where(cond, f(arr), f2(arr))`. _I think_ that the `@jax.jit` will cause `f` not to apply at all to `arr[~cond]` and `f2` not to apply to `arr[cond]`? 
2. (hard) change `@jax.jit` to track unknown shapes:
```python
UNKNOWN_SHAPE_0 = arr[cond].shape  # line 137 
temp1.shape = UNKNOWN_SHAPE_0  # line 137 
UNKNOWN_SHAPE_1 = arr[ncond].shape  # line 153
temp2.shape = UNKNOWN_SHAPE_1  # line 153
out = out.at[ncond].set(temp2)  # line 156; out[ncond] and temp2 have the same shape UNKNOWN_SHAPE_1
out = out.at[cond].set(temp1)  # line 158; out[cond] and temp1 have the same shape UNKNOWN_SHAPE_0
```
In other words, this would require `@jax.jit` to support unknown shapes _as long as they are created and reabsorbed within the same jitted code_.

File: .github/workflows/array_api.yml [hunk 3]
```diff
@@ -16,23 +16,23 @@ env:
...
     -t scipy.optimize.tests.test_bracket
     -t scipy.optimize.tests.test_chandrupatla
     -t scipy.optimize.tests.test_optimize
-    -t scipy.stats
-    -t scipy.ndimage
-    -t scipy.integrate.tests.test_quadrature
     -t scipy.signal.tests.test_signaltools
+    -t scipy.special.tests.test_logsumexp
+    -t scipy.special.tests.test_support_alternative_backends
+    -t scipy.stats
```
Comment (crusaderky) [inline .github/workflows/array_api.yml:35]: just a cosmetic reorder

File: scipy/_lib/tests/test__util.py [hunk 4]
```diff
@@ -347,17 +345,13 @@ def test_contains_nan_with_strings(self):
...
-    @pytest.mark.usefixtures("skip_xp_backends")
     @array_api_compatible
     @pytest.mark.parametrize("nan_policy", ['propagate', 'omit', 'raise'])
     def test_array_api(self, xp, nan_policy):
         rng = np.random.default_rng(932347235892482)
         x0 = rng.random(size=(2, 3, 4))
         x = xp.asarray(x0)
-        x_nan = xp_copy(x, xp=xp)
-        x_nan[1, 2, 1] = np.nan
+        x_nan = xp_at(x)[1, 2, 1].set(np.nan)
```
Comment (lucascolley) [inline scipy/_lib/tests/test__util.py:354]: cosmetic: what do you think about following the pattern used elsewhere and calling this like

```python
import array_api_extra as xpx
...
xpx.at(...)
```

I think that nicely conveys that this is directly from array_api_extra rather than our own custom helper

File: scipy/_lib/_array_api.py [hunk 5]
```diff
@@ -215,12 +216,9 @@ def xp_copy(x: Array, *, xp: ModuleType | None = None) -> Array:
     This copy function does not offer all the semantics of `np.copy`, i.e. the
     `subok` and `order` keywords are not used.
     """
-    # Note: for older NumPy versions, `np.asarray` did not support the `copy` kwarg,
-    # so this uses our other helper `_asarray`.
     if xp is None:
         xp = array_namespace(x)
-
-    return _asarray(x, copy=True, xp=xp)
+    return xp.asarray(x, copy=True)
```
Comment (lucascolley) [inline scipy/_lib/_array_api.py:221]: can we rely on this working for `np<2`? Or are there no codepaths where `xp` is the raw NumPy namespace?

File: scipy/special/tests/test_support_alternative_backends.py [hunk 6]
```diff
@@ -48,8 +48,6 @@ def test_rel_entr_generic(dtype):
 
 @pytest.mark.fail_slow(5)
 @array_api_compatible
-# @pytest.mark.skip_xp_backends('numpy', reason='skip while debugging')
-# @pytest.mark.usefixtures("skip_xp_backends")
```
Comment (lucascolley) [inline scipy/special/tests/test_support_alternative_backends.py:52]: IIRC @mdhaber left this in deliberately as an instant reminder of the syntax (which has changed a few times) if one has to return to debugging here at a later point. But probably fine to remove if you feel strongly.

File: scipy/_lib/_util.py (hunk 2 above)
Comment (crusaderky) [reply scipy/_lib/_util.py:158]: CC @rgommers @jakevdp @pearu

Comment (ev-br) [reply scipy/_lib/_util.py:158]: As a general comment from 30000 feet (am not reviewing this specific PR and do not know enough about jax internals to say anything about specifics):

if a body of this sort of backend-specific workarounds grows beyond a handful of functions to put into array-api-extra, I would suggest to seriously consider using a delegation story [1]. These jax-delegating functions may either live in jax itself or in a separate package, as decided by jax devs and scipy collaborators invested in jax support. Making scipy functions know if they are being jax-jitted or not is just a bit too much. Same for torch.compile or any other existing or future jit frameworks and, in fact, generally the majority of backend-specific code.


[1] For cupy, we make, for instance, `scipy.ndimage.<function>` _delegate_ to `cupyx.scipy.ndimage.<function>` if inputs are cupy arrays. Ditto for `jax.scipy` functions.

Comment (crusaderky) [reply scipy/_lib/_util.py:158]: CC @vfdev-5

File: scipy/_lib/_array_api.py (hunk 5 above)
Comment (crusaderky) [reply scipy/_lib/_array_api.py:221]: I think that `xp` here is `array_api_compat.numpy`, which handles it.
Unit tests will confirm/dismiss my understanding when they finish.

File: scipy/special/tests/test_support_alternative_backends.py (hunk 6 above)
Comment (crusaderky) [reply scipy/special/tests/test_support_alternative_backends.py:52]: I don't feel strongly, but it felt nice to clean up

File: scipy/_lib/_array_api.py (hunk 5 above)
Comment (crusaderky) [reply scipy/_lib/_array_api.py:221]: ...it is not. reverting.

File: scipy/_lib/_util.py (hunk 2 above)
Comment (jakevdp) [reply scipy/_lib/_util.py]: > * (easy-ish?) Reimplement this function in JAX or in array-api-extra and use two different code paths if it is being jitted or not. (_does such a flag exist today?_)

There is no supported way in JAX to check if code is under JIT.

>   If jitted, use `out = jnp.where(cond, f(arr), f2(arr))`. _I think_ that the `@jax.jit` will cause `f` not to apply at all to `arr[~cond]` and `f2` not to apply to `arr[cond]`?

XLA will generally execute both `f` and `f2` on the entire array. This is because on accelerators, there is no efficient way to apply a function to only part of an array. Nevertheless, using `where` would be the recommended way to write this operation in JAX.

> * (hard) change `@jax.jit` to track unknown shapes:

LOL

Comment (crusaderky) [reply scipy/_lib/_util.py]: > There is no supported API in JAX to check if code is under JIT

@vfdev-5 suggested offline:
> if the functions is jitted, then all array-like structures are replaced with a Tracer objects so to know if you are in the eager mode or not, you can do as here: https://github.com/jax-ml/jax/blob/a123d4e39ee67585c0a9dde4694db38ba1aa3e9a/jax/_src/numpy/lax_numpy.py#L6808-L6809

does your comment mean that such a test is disencouraged because it requires the user to test two separate code paths, or that it won't work in some cases?

> LOL

Could you expand? Does it mean "LOL, that's nightmarishly hard don't even bother trying" or "LOL, we've been working on this for ages and we're on track to deliver it already"?

Comment (crusaderky) [reply scipy/_lib/_util.py]: > on accelerators, there is no efficient way to apply a function to only part of an array at dynamically-specified locations.

...from my memory of CUDA 101, you can apply `y = cond ? f(x) : f2(x)` but with the caveat that if cond has disjoint values within the warp (32 contiguous points) you _may_ end up waiting for both f1 and f2?
XREF https://stackoverflow.com/questions/29896422/cuda-avoiding-serial-execution-on-branch-divergence

Comment (jakevdp) [reply scipy/_lib/_util.py]: > does your comment mean that such a test is disencouraged because it requires the user to test two separate code paths, or that it won't work in some cases?

There is no public API to check whether you're under JIT, because doing so is a very easy way to inadvertently introduce subtle bugs into your code. It is of course possible to use non-public implementation details to infer whether you're in JIT, and that might work until those implementation details change. That's not a path I'd suggest for a project as important as SciPy.

> from my memory of CUDA 101, you can apply `y = cond ? f1(x) : f2(x)` but with the caveat that if cond has disjoint values within the warp (32 contiguous points) you _may_ end up waiting for both f1 and f2?

Yes, that's the case for a scalar `cond`, and JAX provides this API via `jax.lax.cond`. The context here if I understand correctly is a vector `cond`, which is a different situation.

> Could you expand?

Sorry, LOL probably wasn't that helpful. There have been several attempts at introducing some form of dynamic shapes into JAX over the years, but all have run into fundamental difficulties, not least of which is the non-existence of a suitable dynamic-shape-aware compiler. It may happen some day, but "hard" is an understatement 😁

Comment (crusaderky) [reply scipy/_lib/_util.py]: > Yes, that's the case for a scalar cond, and JAX provides this API via [jax.lax.cond](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.cond.html). The context here if I understand correctly is a vector cond, which is a different situation.

My snippet was in the context of an embarassingly-parallel CUDA kernel, where x, y, and cond are single points out of vectors

Comment (crusaderky) [reply scipy/_lib/_util.py]: > There is no public API to check whether you're under JIT, because doing so is a very easy way to inadvertently introduce subtle bugs into your code. It is of course possible to use non-public implementation details to infer whether you're in JIT, and that might work until those implementation details change. That's not a path I'd suggest for a project as important as SciPy.

Would you agree that `_lazy_where` (I dislike the name though; I think `apply_where` would be better) is a very generic and powerful function and it should be implemented in JAX with two separate code paths for jit vs. non-jit, plus a wrapper in `array-api-extra` to cover all other array libraries?

Comment (jakevdp) [reply scipy/_lib/_util.py]: My understanding is that embarrassingly parallel in CUDA implies vectorization. In JAX that’s certainly the case

Comment (crusaderky) [reply scipy/_lib/_util.py]: Alternatively, we could expand `jnp.at[ ].apply()` to work with any jit-table callable instead of just ufuncs?

**Comment by crusaderky** [comment]
Unit test failures are unrelated.
This is ready for final review.
For merge, it will need to wait for the next releases of `array-api-compat` and `array-api-extra`.

**Review by ev-br — COMMENTED** [review]
For `_lazywhere` specifically, it seems to make sense to move this implementation to 
https://github.com/data-apis/array-api-extra/issues/14 and reuse across the scipy codebase after array-api-extra cut the release.

File: scipy/_lib/_util.py [hunk 7]
```diff
@@ -153,9 +154,9 @@ def _lazywhere(cond, arrays, f, fillvalue=None, f2=None):
         temp2 = xp.asarray(f2(*(arr[ncond] for arr in arrays)))
         dtype = xp.result_type(temp1, temp2)
         out = xp.empty(cond.shape, dtype=dtype)
-        out[ncond] = temp2
+        out = xpx.at(out, ncond).set(temp2, copy=None)
```
[the remark this answers is not in the document]
Comment (pearu) [reply scipy/_lib/_util.py:157]: See https://github.com/data-apis/array-api/issues/845#issuecomment-2388436813. In short, `__setitem__` is an in-place-only operation while `.at(...).set(..)` ought to create a new array.

[the remark this answers is not in the document]
Comment (pearu) [reply scipy/_lib/_util.py:157]: Although, Python allows `__setitem__` to return values so that if one defines
```python
def __setitem__(self, key, item):
    return xpx.at(self, key).set(item, copy=None)
```
then the mouthful expression could be written as
```python
out = out.__setitem__(ncond, temp2)
```

[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py:157]: > If not, please detail the reasons in code comments here

The only purpose of `xpx.at` is to allow update operations for immutable libraries. That should be explained clearly in the docstring of `xpx.at` itself. If you think that https://github.com/data-apis/array-api-extra/pull/53/ lacks clarity in its documentation, I'm happy to improve it there. I think it would be very redundant and verbose to repeat it every time the function is invoked in scipy (there will be _lots_; this is the first of many more PRs).

[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py:157]: > Although, Python allows `__setitem__` to return values so that if one defines
> 
> ```python
> def __setitem__(self, key, item):
>     return xpx.at(self, key).set(item, copy=None)
> ```
> 
> then the mouthful expression could be written as
> 
> ```python
> out = out.__setitem__(ncond, temp2)
> ```

There would be no way for the library implementing the `__setitem__` method to know if the user called the syntax you propose or if they just wrote `out[ncond] = temp2`, thus silently losing the output.
```python
In [2]: class C:
   ...:     def __init__(self, a):
   ...:         self.a = a
   ...:     def __setitem__(self, idx, value):
   ...:         a = self.a[:]
   ...:         a[idx] = value
   ...:         return C(a)
   ...: 

In [3]: c = C([1,2,3])

In [4]: c[0] = 4

In [5]: c.a
Out[5]: [1, 2, 3]

In [6]: c = c.__setitem__(0, 4)

In [7]: c.a
Out[7]: [4, 2, 3]
```

[the remark this answers is not in the document]
Comment (pearu) [reply scipy/_lib/_util.py:157]: > There would be no way for the library implementing the __setitem__ method to know if the user called the syntax you propose or if they just wrote out[ncond] = temp2, thus silently losing the output.

Exactly. The library has to be consistent with using `out = out.__setitem__(...)` or using `out = xpx.at(...).set(...)`. These two usage cases are equivalent in the sense that once it uses  `out[...] = ...`, the output is silently lost (unless `__setitem__` raises an exception that is the current JAX behavior).

I am +0 for using `out = out.__setitem__(...)` just because I can also feel the slight pain of reading/writing `out = xpx.at(...).set(...)` but on the other hand, JAX has established this and `__setitem__` behavior for a long time ago.

[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py:157]: To me `out = out.__setitem__(...)` is a strong -1, because novice JAX users would routinely lose extensive amount of time and energy trying to figure out why algorithms they just ported verbatim from numpy won't work with jnp. The informative exception they're receiving now is a reasonable way to handle JAX's immutability IMHO.

[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py:157]: I would not mind however convenience aliases e.g.
```python
def set_at(x, idx, value):
    return at(x, idx).set(value)
```
however such discussion should be held at https://github.com/data-apis/array-api-extra/pull/53.

[the remark this answers is not in the document]
Comment (ev-br) [reply scipy/_lib/_util.py:157]: > ... function is invoked in scipy (there will be lots; this is the first of many more PRs).

Was there a mailing list (== discourse) discussion on this? I must have missed it, mind dropping a link here?

> In short, __setitem__ is an in-place-only operation while .at(...).set(..) ought to create a new array.

Aha. 
1) What are the memory implications then? Boolean or slice indexing in numpy or torch is `O(r.h.s. size)` complexity and no extra allocations, what about the out-of-place version?
2) What are implications for `torch.compile`? I vaguely remember that tracking the identity of an object is sometimes crucial to avoid graph breaks or IIRC outright wrong results.

[the remark this answers is not in the document]
Comment (pearu) [reply scipy/_lib/_util.py:157]: > > ... function is invoked in scipy (there will be lots; this is the first of many more PRs).
> 
> Was there a mailing list (== discourse) discussion on this? I must have missed it, mind dropping a link here?

No, AFAIK. Closest to this is the discussion held in https://github.com/data-apis/array-api/issues/845

> > In short, `__setitem__` is an in-place-only operation while `.at(...).set(..)` ought to create a new array.
> 
> Aha.
> 
> 1. What are the memory implications then?

None. `out.at(...).set(..)` is equivalent to `out[..] = ..` for libraries that support mutable arrays. JAX JIT compilation optimizes expressions like `out.at(...).set(..)` before any memory implications, roughly speaking.

> Boolean or slice indexing in numpy or torch is `O(r.h.s. size)` complexity and no extra allocations, what about the out-of-place version?

By "out-of-place", do you mean using `__getitem__`? If so, it is OT AFAICT.

> 2. What are implications for `torch.compile`? I vaguely remember that tracking the identity of an object is sometimes crucial to avoid graph breaks or IIRC outright wrong results.

From reading https://github.com/data-apis/array-api-extra/pull/53, `torch.compile` would see `out.at(...).set(..)` as normal `out[..] = ..`.

[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py:157]: > None. out.at(...).set(..) is equivalent to out[..] = .. for libraries that support mutable arrays.

You have to explicitly call `set(..., copy=None)`. By default (copy=True) everything is deep-copied ahead of the change. This is a conscious design choice of safety over speed by default.

[the remark this answers is not in the document]
Comment (ev-br) [reply scipy/_lib/_util.py:157]: >>> ... function is invoked in scipy (there will be lots; this is the first of many more PRs).

>> Was there a mailing list (== discourse) discussion on this? I must have missed it, mind dropping a link here?

> No, AFAIK. Closest to this is the discussion held in https://github.com/data-apis/array-api/issues/845

So it's about time to start one, it seems :-).

-------------------------------

>> What are the memory implications then?

> None. out.at(...).set(..) is equivalent to out[..] = .. for libraries that support mutable arrays. JAX JIT compilation optimizes expressions like out.at(...).set(..) before any memory implications, roughly speaking.

Thanks. What about other libraries though. With, say, numpy, does `out = out.at(...).set(...)` allocate a new array for the right-hand side?
My concern is whether catering for jax's immutability does not pessimize performance for other backends.
 

-----------------------------------------

> From reading https://github.com/data-apis/array-api-extra/pull/53, torch.compile would see out.at(...).set(..) as normal out[..] = ...

Would actually be nice to double-check for things like
https://github.com/pytorch/pytorch/issues/113539 and https://github.com/pytorch/pytorch/pull/113931

**Comment by andyfaff** [comment]
> out.at(...).set(..) is equivalent to out[..] = .. for libraries that support mutable arrays

> My concern is whether catering for jax's immutability does not pessimize performance for other backends.

I second that. JAX support shouldn't be at the cost of degrading memory/CPU/GPU load.

Furthermore, re-writing code for jax compliance may be quite a burden. For example, (IMO) just the `out.at(...).set(..)` would make the entire codebase harder to follow from a cognitive aspect.

File: scipy/_lib/_util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py:157]: > Thanks. What about other libraries though. With, say, numpy, does out = out.at(...).set(...) allocate a new array for the right-hand side?

This question is answered in the xpx.at documentation: https://github.com/data-apis/array-api-extra/pull/53/files#diff-ac20f7d69cdfe2230e1ccd57f43df58c0126f2b21b0013eececf8aeab0f07179R609-R708

> My concern is whether catering for jax's immutability does not pessimize performance for other backends.

It does not.

> Would actually be nice to double-check for things like
https://github.com/pytorch/pytorch/issues/113539 and https://github.com/pytorch/pytorch/pull/113931

Neither in-place resize nor in-place sort are part of the array API specifications, so as long as scipy remains array API compliant they will only become a problem in torch (and jax, and sparse) when and if they're added to the standard.

[the remark this answers is not in the document]
Comment (pearu) [reply scipy/_lib/_util.py:157]: > Thanks. What about other libraries though. With, say, numpy, does out = out.at(...).set(...) allocate a new array for the right-hand side?

No, unless the array `out` is immutable or `.set(..., copy=True)` is used.

> My concern is whether catering for jax's immutability does not pessimize performance for other backends.

@crusaderky , do we have tests for mutable arrays that are
```python
out_set = out.at(...).set(..., copy=None)
assert out_set is out
```
? If not, I suggest introducing such tests to relieve concerns about possible copies from using the `out.at(...).set(..., copy=None)` pattern.

**Comment by crusaderky** [comment]
> Furthermore, re-writing code for jax compliance may be quite a burden.

This is for forward compliance with all read-only array API backends. Namely, to extend support to pydata sparse one would just need to implement `.at[...]` or similar in sparse and then call it from `xpx.at`.

**Comment by lucascolley** [comment]
> > Furthermore, re-writing code for jax compliance may be quite a burden.
> This is for forward compliance with all read-only array API backends. Namely, to extend support to pydata sparse one would just need to implement .at[...] or similar in sparse and then call it from xpx.at.

I agree, but this discussion probably belongs on https://discuss.scientific-python.org/c/contributor/scipy/32 rather than here.

**Comment by ev-br** [comment]
To land some form of this in scipy, I think it really needs an RFC-type proposal.
It should address at least these concerns (am listing mine just to give a non-exhaustive list):

- the proposed syntax
- an estimate of the blast radius for the scipy code
- why does it require scipy code changes instead of delegating to something specialized
- what are the positives, actually; what do we get for the readability hit
- the (lack of) effect for backends with mutable arrays (tests that Pearu suggested above would go a long way towards this)
- how this interacts with compiled code (lapack, our custom things in C or Cython)
- how this interacts with JAX JIT (makes it possible to JIT a scipy function with observable benefits? a demo would go a long way)
- how this interacts with non-JAX JIT, e.g. torch.compile which tracks an object's identity (again, a demo would be a very good to have, so that we can ask seasoned pytorch/dynamo devs to vet it)
- it would really help to see a worked example. `_lazywhere` is a utility function and like I was saying above it might be even better to move to `xpx`. I'd suggest to pick some public function, convert it, and use the result to demonstrate the benefits.

**Comment by rgommers** [comment]
Hey all, a quick comment while I'm catching up from the end-of-2024-rush and holiday. This PR follows up on gh-20085 which introduced JAX support, and in particular the discussion on in-place operations at the end of that PR (discussion mostly between (@lucascolley, @mdhaber, @jakevdp and me there). I prototyped this `at` usage already there - @crusaderky did a much better job now than what I did there, and my initial impression is that this is pretty clean. Many of the questions being brought up now were already discussed or answered there, including the clearly important one of "what is the benefit". [This comment](https://github.com/scipy/scipy/pull/20085#issuecomment-2119307509) contains a demo/benchmark, copying the relevant part here:

_I could get `cluster.whiten` to work with `jax.jit` on CPU and GPU with some minor tweaks. And it helps performance:_
```
    whiten_jit(face).block_until_ready()  # do the JIT compilation
    face_gpu = face  # JAX defaults to GPU if that is available
    face_cpu = jax.device_put(face, jax.devices('cpu')[0])
    face_np = np.asarray(face)

    %timeit cluster.vq.whiten(face_np)   #  22 ms
    %timeit cluster.vq.whiten(face_cpu)  #   6 ms
    %timeit whiten_jit(face_cpu)         #   3 ms
    %timeit cluster.vq.whiten(face_gpu)  # 700 us
    %timeit whiten_jit(face_gpu)         # 275 us
```

I'll try to review this asap. Let's please avoid giving @crusaderky a lot of extra homework right this minute until I've done that.

File: scipy/special/tests/test_support_alternative_backends.py (hunk 6 above)
Comment (crusaderky) [reply scipy/special/tests/test_support_alternative_backends.py]: Reverted and moved to a later PR specific to scipy.special

File: scipy/_lib/_util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py]: @pearu, yes, https://github.com/data-apis/array-api-extra/pull/53 extensively tests that copy=None maps to True for jax and read-only numpy and to False for everything else.

[the remark this answers is not in the document]
Comment (crusaderky) [reply scipy/_lib/_util.py]: Changed default value of copy= from True to False following https://github.com/data-apis/array-api-extra/pull/53#pullrequestreview-2528198254

**Comment by crusaderky** [comment]
array-api-extra has been released (many thanks @lucascolley!!!) and this PR is ready for final review.
After it's merged, I'll start working on moving _lazywhere to array-api-extra, where it will encapsulate a special case path that lets it work with jitted jax.

**Review by lucascolley — APPROVED** [review]

**Review by pearu — APPROVED** [review]
LGTM! The default `copy=None` in `at.set` method makes sense.
Thanks, @crusaderky!

**Review by rgommers — APPROVED** [review]
This looks quite clean now, thanks @crusaderky & all reviewers. I think we can go ahead and merge this. Longer story below to address some of the big-picture review comments that don't relate only to this PR.

* * *

All the questions above are very much valid, but they're just slightly out of place on this pretty small PR I think. My view is that we had an RFC at the beginning for array types support, then have been doing a lot of work under an experimental label (like this PR), and we need an assessment at the end of some topic and the overall effort to decide that something works for us and/or can be moved from experimental to public/supported status. For JAX, we should do this exercise at the end of steps 1-7 below (with sanity checks as we go of course): we expect it to work, and the benefits of supporting JAX to be worth the costs. Those include (a) significant enough performance gains of `jax.jit` vs. `numpy` on CPU and GPU, (b) enough of SciPy's functionality/submodules able to have that support, (c) no negative impact on support for other libraries, and(d)  the overall maintainability impact of JAX support looking acceptable.

The full set of changes that I think are needed for solid JAX support, from beginning to end, are (not necessarily exactly in order):
1. The initial work in gh-20085 that introduced eager JAX support (done)
2. The changes to test skip syntax that that helped trigger (done)
3. This PR for dealing with in-place operations with integer indexing
4. A similar change for dealing with in-place operations with boolean masks
5. A way to avoid `jax.jit` choking on input argument validation (xref https://github.com/data-apis/array-api-compat/issues/225)
6. Making code that calls into compiled code in SciPy work with `jax.jit` (doable as long as shape and dtype of the returned array from the compiled function is predictable, probably with `jax.pure_callback`)
7. Adding a CI job that verifies that `jax.jit` works on SciPy functionality supporting it.

What I think we should do is make these changes as quickly as we can on *a representative subset of SciPy only*, to avoid having a large amount of churn for one of the steps 3-6, across the whole code base and then figuring out that the end goal is somehow blocked by an unforeseen issue. After discussion with @crusaderky I think `cluster` plus the part of `special` that already delegates to JAX for matching functions is suitable for this. 

We can definitely open a new issue for JAX support - I'll take that action (it's basically a more extensive version of this comment, with some cross-linking). The key discussion preceding this PR having been on another PR rather than an issue is slightly suboptimal but doesn't seem like a mortal sin to me, lots of other array API work and other feature additions don't have an issue for every PR/topic.

File: scipy/_lib/_util.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply scipy/_lib/_util.py]: This comment can be resolved now. After the change to `copy=None`, the change from `x[idx] += y` to `x = xpx.at(x)[idx].add(y)` is mechanical and `at` does not make any copies. The `at` docstring is pretty clear and addresses all this, with usage examples.

**Comment by lucascolley** [comment]
thanks all 😄

**Comment by rgommers** [comment]
> For JAX, we should do this exercise at the end of steps 1-7 below

For clarity: the approach I sketched of working out the design and doing a cost-benefit analysis before declaring anything public and stable is not specific to JAX, it applies to all array types support activity.

**Comment by rgommers** [comment]
Opened a tracking issue: gh-22246
